src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs

```diff
@@ -51,73 +46,23 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
                 ? _symbolSearchService ?? solution.Workspace.Services.GetService<ISymbolSearchService>()
                 : null;
 
-            var packageSources = symbolSearchService != null && searchNuGetPackages
-                ? GetPackageSources(document)
+            var installerService = GetPackageInstallerService(document);
+            var packageSources = searchNuGetPackages && symbolSearchService != null && installerService != null
+                ? installerService.PackageSources
                 : ImmutableArray<PackageSource>.Empty;
 
-            // We might have multiple different diagnostics covering the same span.  Have to
-            // process them all as we might produce different fixes for each diagnostic.
-
-            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
-            var placeSystemNamespaceFirst = documentOptions.GetOption(GenerationOptions.PlaceSystemNamespaceFirst);
+            var fixesForDiagnostic = await addImportService
+                .GetFixesForDiagnosticsAsync(document, span, diagnostics, symbolSearchService, searchReferenceAssemblies, packageSources, cancellationToken).ConfigureAwait(false);
 
-            foreach (var diagnostic in context.Diagnostics)
+            foreach (var (diagnostic, fixes) in fixesForDiagnostic)
             {
-                var fixes = await addImportService.GetFixesAsync(
-                    document, span, diagnostic.Id, placeSystemNamespaceFirst,
-                    symbolSearchService, searchReferenceAssemblies, 
-                    packageSources, cancellationToken).ConfigureAwait(false);
-
-                var codeActions = ArrayBuilder<CodeAction>.GetInstance();
-
-                foreach (var fix in fixes)
-                {
-                    var codeAction = TryCreateCodeAction(document, fix);
-                    codeActions.AddIfNotNull(codeAction);
-
-                    if (codeActions.Count >= MaxResults)
-                    {
-                        break;
-                    }
-                }
-
+                // Limit the results returned since this will be displayed to the user
+                var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, installerService, limitResults: true);
```

Comment:
instead of 'limitResults', consider passing in the max amount that should be returned.